Model/Indexer/AbstractIndexer.php

```diff
@@ -0,0 +1,198 @@
+
+    /** @var NostoHelperScope */
+    private $nostoHelperScope;
+
+    /** @var NostoLogger */
+    public $nostoLogger;
+
+    /** @var ProcessManager */
+    private $processManager;
+
+    /** @var DimensionProviderInterface */
+    private $dimensionProvider;
+
+    /**
+     * AbstractIndexer constructor.
+     * @param NostoHelperAccount $nostoHelperAccount
+     * @param NostoHelperScope $nostoHelperScope
+     * @param NostoLogger $nostoLogger
+     * @param DimensionProviderInterface $dimensionProvider
+     * @param ProcessManager|null $processManager
+     */
+    public function __construct(
+        NostoHelperAccount $nostoHelperAccount,
+        NostoHelperScope $nostoHelperScope,
+        NostoLogger $nostoLogger,
+        DimensionProviderInterface $dimensionProvider,
+        ProcessManager $processManager = null
+    ) {
+        $this->nostoHelperAccount = $nostoHelperAccount;
+        $this->nostoHelperScope = $nostoHelperScope;
+        $this->nostoLogger = $nostoLogger;
+        $this->dimensionProvider = $dimensionProvider;
+        $this->processManager = $processManager;
+    }
+
+    /**
+     * Get ModeSwitcher class to later get the indexer mode
+     *
+     * @return ModeSwitcherInterface
+     */
+    abstract public function getModeSwitcher(): ModeSwitcherInterface;
+
+    /**
+     * Implement logic of single store indexing
+     *
+     * @param Store $store
+     * @param array $ids
+     */
+    abstract public function doIndex(Store $store, array $ids = []);
+
+    /**
+     * @return string
+     */
+    abstract public function getIndexerId(): string ;
+
+    /**
+     * @param array $ids
+     * @suppress PhanTypeMismatchArgument
+     */
+    public function doWork(array $ids = [])
+    {
+        $userFunctions = [];
+
+        switch ($this->getModeSwitcher()->getMode()) {
+            case DimensionModeConfiguration::DIMENSION_NONE:
+                $userFunctions[] = function () use ($ids) {
+                    $storesWithNosto = $this->nostoHelperAccount->getStoresWithNosto();
+                    foreach ($storesWithNosto as $store) {
+                        $this->doIndex($store, $ids);
+                    }
+                };
+                break;
+            case DimensionModeConfiguration::DIMENSION_STORE:
+                /** @var Dimension[] $dimension  */
+                foreach ($this->dimensionProvider->getIterator() as $dimension) {
+                    /** @suppress PhanTypeMismatchArgument */
+                    $userFunctions[] = function () use ($dimension, $ids) {
+                        $this->executeByDimensions($dimension, new ArrayIterator($ids));
+                    };
+                }
+                break;
+            default:
+                throw new UnexpectedValueException("Undefined dimension mode.");
+        }
+        /** @var Traversable $userFunctions  */
+        $this->getProcessManager()->execute($userFunctions);
```

Comment:
Why is this being at the end? Shouldn't this be before the break in the L135?

Reply:
Because it is common for both case to execute the array of functions.

Reply:
When the indexer mode is `none`, we should just run the functions sequentially. When the mode is `store`, we build the list of functions and then call the process-manager. What am i missing?